### PR TITLE
Test cpp-simple image

### LINKED ISSUE #3480: Add e2e test for CPP example game server
**Is your feature request related to a problem? Please describe.**
Right now it isn't clear how to verify changes to the [CPP example](https://github.com/googleforgames/agones/tree/main/examples/cpp-simple) are correct and still work properly. 

**Describe the solution you'd like**
A basic smoke test that verifies that the game server starts and becomes ready. 

**Describe alternatives you've considered**
Write documentation to do manual testing.

### PR DESCRIPTION
**Pull request by Kalaiselvi84**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

/kind cleanup
> /kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #3480 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by markmandel** [comment]
Looks like a flake downloading Helm.

```
Removing intermediate container c261418f2a21
 ---> b66c8411921f
Step 13/19 : RUN curl -L  ${HELM_URL} > /tmp/helm.tar.gz     && tar -zxvf /tmp/helm.tar.gz -C /tmp     && mv /tmp/linux-amd64/helm /usr/local/bin/helm     && chmod go+rx /usr/local/bin/helm     && rm /tmp/helm.tar.gz && rm -rf /tmp/linux-amd64
 ---> Running in 77bdc37bcf0e
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

curl: (28) Operation timed out after 300204 milliseconds with 0 out of 0 bytes received
The command '/bin/sh -c curl -L  ${HELM_URL} > /tmp/helm.tar.gz     && tar -zxvf /tmp/helm.tar.gz -C /tmp     && mv /tmp/linux-amd64/helm /usr/local/bin/helm     && chmod go+rx /usr/local/bin/helm
```

My suggestion - fix the conflict and re-run when ready.

**Comment by markmandel** [comment]
oooh, that's a new flake:

https://console.cloud.google.com/cloud-build/builds/8dbdc9ec-9708-4df4-86e6-e53ee707926d;step=20?e=13803378&mods=logs_tg_prod&project=agones-images

```
==================
WARNING: DATA RACE
Read at 0x00c00168a510 by goroutine 860:
  reflect.maplen()
      /usr/local/go/src/runtime/map.go:1394 +0x0
  reflect.Value.lenNonSlice()
      /usr/local/go/src/reflect/value.go:1704 +0x26d
  reflect.Value.Len()
      /usr/local/go/src/reflect/value.go:1693 +0x3e6
  encoding/json.mapEncoder.encode()
      /usr/local/go/src/encoding/json/encode.go:795 +0x3cb
  encoding/json.mapEncoder.encode-fm()
      <autogenerated>:1 +0x90
  encoding/json.(*encodeState).reflectValue()
      /usr/local/go/src/encoding/json/encode.go:358 +0x88
  encoding/json.interfaceEncoder()
      /usr/local/go/src/encoding/json/encode.go:714 +0x17b
  encoding/json.mapEncoder.encode()
      /usr/local/go/src/encoding/json/encode.go:812 +0x765
  encoding/json.mapEncoder.encode-fm()
      <autogenerated>:1 +0x90
  encoding/json.(*encodeState).reflectValue()
      /usr/local/go/src/encoding/json/encode.go:358 +0x88
  encoding/json.(*encodeState).marshal()
      /usr/local/go/src/encoding/json/encode.go:330 +0x224
  encoding/json.(*Encoder).Encode()
      /usr/local/go/src/encoding/json/stream.go:209 +0x146
  github.com/sirupsen/logrus.(*JSONFormatter).Format()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/json_formatter.go:123 +0xf91
  github.com/sirupsen/logrus.(*Entry).write()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/entry.go:289 +0x133
  github.com/sirupsen/logrus.(*Entry).log()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/entry.go:252 +0x9b2
  github.com/sirupsen/logrus.(*Entry).Log()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/entry.go:304 +0x8b
  github.com/sirupsen/logrus.(*Entry).Debug()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/entry.go:313 +0x22d
  agones.dev/agones/pkg/sdkserver.(*SDKServer).updateList()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:1178 +0xb5
  agones.dev/agones/pkg/sdkserver.(*SDKServer).syncGameServer()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:349 +0xe6
  agones.dev/agones/pkg/sdkserver.(*SDKServer).syncGameServer-fm()
      <autogenerated>:1 +0x6d
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).processNextWorkItem()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:182 +0x366
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).runWorker()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:158 +0x4c
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run.func1()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:217 +0x14
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x48
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xce
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x10d
  k8s.io/apimachinery/pkg/util/wait.Until()
      /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161 +0x116
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:217 +0x93
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).Run.func1()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:206 +0x58

Previous write at 0x00c00168a510 by goroutine 842:
  runtime.mapassign_faststr()
      /usr/local/go/src/runtime/map_faststr.go:203 +0x0
  agones.dev/agones/pkg/sdkserver.(*SDKServer).AddListValue()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:1122 +0xad0
  agones.dev/agones/pkg/sdkserver.TestSDKServerAddListValue.func1()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1441 +0xcc7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Goroutine 860 (running) created at:
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).Run()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:206 +0x1a6
  agones.dev/agones/pkg/sdkserver.(*SDKServer).Run()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:293 +0xf0d
  agones.dev/agones/pkg/sdkserver.TestSDKServerAddListValue.func1.3()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1428 +0x6e

Goroutine 842 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  agones.dev/agones/pkg/sdkserver.TestSDKServerAddListValue()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1387 +0x22c4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c000b31348 by goroutine 860:
  reflect.typedmemmove()
      /usr/local/go/src/runtime/mbarrier.go:178 +0x0
  reflect.copyVal()
      /usr/local/go/src/reflect/value.go:1978 +0x6e
  reflect.(*MapIter).Value()
      /usr/local/go/src/reflect/value.go:1880 +0x111
  encoding/json.mapEncoder.encode()
      /usr/local/go/src/encoding/json/encode.go:799 +0x5cb
  encoding/json.mapEncoder.encode-fm()
      <autogenerated>:1 +0x90
  encoding/json.(*encodeState).reflectValue()
      /usr/local/go/src/encoding/json/encode.go:358 +0x88
  encoding/json.interfaceEncoder()
      /usr/local/go/src/encoding/json/encode.go:714 +0x17b
  encoding/json.mapEncoder.encode()
      /usr/local/go/src/encoding/json/encode.go:812 +0x765
  encoding/json.mapEncoder.encode-fm()
      <autogenerated>:1 +0x90
  encoding/json.(*encodeState).reflectValue()
      /usr/local/go/src/encoding/json/encode.go:358 +0x88
  encoding/json.(*encodeState).marshal()
      /usr/local/go/src/encoding/json/encode.go:330 +0x224
  encoding/json.(*Encoder).Encode()
      /usr/local/go/src/encoding/json/stream.go:209 +0x146
  github.com/sirupsen/logrus.(*JSONFormatter).Format()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/json_formatter.go:123 +0xf91
  github.com/sirupsen/logrus.(*Entry).write()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/entry.go:289 +0x133
  github.com/sirupsen/logrus.(*Entry).log()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/entry.go:252 +0x9b2
  github.com/sirupsen/logrus.(*Entry).Log()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/entry.go:304 +0x8b
  github.com/sirupsen/logrus.(*Entry).Debug()
      /go/src/agones.dev/agones/vendor/github.com/sirupsen/logrus/entry.go:313 +0x22d
  agones.dev/agones/pkg/sdkserver.(*SDKServer).updateList()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:1178 +0xb5
  agones.dev/agones/pkg/sdkserver.(*SDKServer).syncGameServer()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:349 +0xe6
  agones.dev/agones/pkg/sdkserver.(*SDKServer).syncGameServer-fm()
      <autogenerated>:1 +0x6d
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).processNextWorkItem()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:182 +0x366
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).runWorker()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:158 +0x4c
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run.func1()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:217 +0x14
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x48
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xce
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x10d
  k8s.io/apimachinery/pkg/util/wait.Until()
      /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161 +0x116
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:217 +0x93
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).Run.func1()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:206 +0x58

Previous write at 0x00c000b31348 by goroutine 842:
  agones.dev/agones/pkg/sdkserver.(*SDKServer).AddListValue()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:1122 +0xae4
  agones.dev/agones/pkg/sdkserver.TestSDKServerAddListValue.func1()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1441 +0xcc7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Goroutine 860 (running) created at:
  agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).Run()
      /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:206 +0x1a6
  agones.dev/agones/pkg/sdkserver.(*SDKServer).Run()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:293 +0xf0d
  agones.dev/agones/pkg/sdkserver.TestSDKServerAddListValue.func1.3()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1428 +0x6e

Goroutine 842 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  agones.dev/agones/pkg/sdkserver.TestSDKServerAddListValue()
      /go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1387 +0x22c4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47
==================
```

@igooch looks like this is happening in List functions, does anything immediately jump out at you?

**Comment by igooch** [comment]
> oooh, that's a new flake:
> 
> https://console.cloud.google.com/cloud-build/builds/8dbdc9ec-9708-4df4-86e6-e53ee707926d;step=20?e=13803378&mods=logs_tg_prod&project=agones-images

> @igooch looks like this is happening in List functions/tests, does anything immediately jump out at you?

Looks like there's a flake, running `~/agones/pkg/sdk$ go test -timeout 300s -run ^TestSDKServerAddListValue$ agones.dev/agones/pkg/sdkserver -race -count=10` passes, but `~/agones/pkg/sdk$ go test -timeout 300s -run ^TestSDKServerAddListValue$ agones.dev/agones/pkg/sdkserver -race -count=100` fails. Nothing jumps out immediately at me though, I'll need to take a look.

File: test/e2e/examples_test.go [hunk 1]
```diff
@@ -64,3 +64,38 @@ func TestSuperTuxKartGameServerReady(t *testing.T) {
...
+				ContainerPort: 7654,
+				Protocol:      corev1.ProtocolUDP,
+			}},
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            "cpp-simple",
+							Image:           "us-docker.pkg.dev/agones-images/examples/cpp-simple-server:0.15",
+							ImagePullPolicy: corev1.PullAlways,
```
Comment (Kalaiselvi84) [inline test/e2e/examples_test.go:87]: Will remove ImagePullPolicy from this test

**Comment by Kalaiselvi84** [comment]
@markmandel Are we good to merge this PR?

**Review by markmandel — APPROVED** [review]

**Pull request merged**